coreneuron/io/core2nrn_data_return.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright (c) 2020, Blue Brain Project
+All rights reserved.
+     
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "coreneuron/coreneuron.hpp"
+#include "coreneuron/io/nrn2core_direct.h"
+#include "coreneuron/sim/multicore.hpp"
+#include "coreneuron/nrniv/nrniv_decl.h"
+#include "coreneuron/io/core2nrn_data_return.hpp"
+
+/** @brief, Information from NEURON to help with copying data to NEURON.
+ *  Info for copying voltage, i_membrane_, and mechanism data.
+ *  See implementaton in
+ *  nrn/src/nrniv/nrnbbcore_write.cpp:nrnthreads_type_return.
+ *  Return is size of either the returned data pointer or the number
+ *  of pointers in mdata. tid is the thread index.
+ */
+size_t (*nrn2core_type_return_)(int type, int tid,
+  double*& data, double**& mdata);
+
+namespace coreneuron {
+
+/** @brief permuted array copied to unpermuted array
+ *  If permute is NULL then just a copy
+ */
+static void inverse_permute_copy(size_t n, double* permuted_src, double* dest,
+                                 int* permute)
```

Comment:
Just a remark for the overall code : we typically use `const` for  pointer parameters for the reasons [discussed here](https://stackoverflow.com/questions/7715371/whats-the-point-of-const-pointers).

Reply:
Thanks. Does the fact that permuted_src and permute elements do not change mean that I should
```
static void inverse_permute_copy(size_t n, const double* const permuted_src, double* const dest,
                                 const int* const permute)
```
I have to admit I never thought it was obvious which const meant what.

Reply:
> Thanks. Does the fact that permuted_src and permute elements do not change mean that I should

> static void inverse_permute_copy(size_t n, const double* const permuted_src, double* const dest,
>                                 const int* const permute)

In this case,

* `permuted_src` and `permute` are read-only and remain constant. So it's important to declare them with `double* const ` and `int* const`
* additionally, when we say `const int* permute`, this means we can't change permute itself with `permute = new_ptr`. This is fine as well. But just want to mention that as pointer is passed by value (and not reference),  even if pointer is changed, it doesn't affect the caller. 


> static void soa2aos_inverse_permute_copy(size_t n, int sz, int stride,
  const double* const src, double* const * const dest, const int* const permute)

In this case we are changing values inside dest[i][j] and hence it can't be `* const dest` ?

Reply:
To take care of with refactoring